src/cart/crt_hg_proc.c

```diff
@@ -254,6 +254,52 @@ crt_proc_d_iov_t(crt_proc_t proc, crt_proc_op_t proc_op, d_iov_t *div)
 	return rc;
 }
 
+int
+crt_proc_d_sg_list_t(crt_proc_t proc, crt_proc_op_t proc_op, d_sg_list_t *p)
+{
+	int		i;
+	int		rc;
+
+	if (FREEING(proc_op)) {
+		/* NB: don't need free in crt_proc_d_iov_t() */
+		D_FREE(p->sg_iovs);
+		return 0;
+	}
+
+	rc = crt_proc_uint32_t(proc, proc_op, &p->sg_nr);
+	if (unlikely(rc))
+		return rc;
+
+	rc = crt_proc_uint32_t(proc, proc_op, &p->sg_nr_out);
+	if (unlikely(rc))
+		return rc;
+
+	if (p->sg_nr == 0)
+		return 0;
+
+	switch (proc_op) {
+	case CRT_PROC_DECODE:
+		D_ALLOC_ARRAY(p->sg_iovs, p->sg_nr);
+		if (p->sg_iovs == NULL)
+			return -DER_NOMEM;
+		/* fall through to fill sg_iovs */
+	case CRT_PROC_ENCODE:
+		for (i = 0; i < p->sg_nr; i++) {
+			rc = crt_proc_d_iov_t(proc, proc_op, &p->sg_iovs[i]);
+			if (unlikely(rc)) {
+				if (DECODING(proc_op))
+					D_FREE(p->sg_iovs);
+				return rc;
+			}
+		}
+		break;
+	default:
+		return -DER_INVAL;
+	}
```

Comment:
I'd missed this switch statement was falling through although now I see there's even a comment.

Couldn't the same be achieved by a simple if statement?
```suggestion
	if (DECODING(proc_op)) {
		D_ALLOC_ARRAY(p->sg_iovs, p->sg_nr);
		if (p->sg_iovs == NULL)
			return -DER_NOMEM;
	}
	for (i = 0; i < p->sg_nr; i++) {
		rc = crt_proc_d_iov_t(proc, proc_op, &p->sg_iovs[i]);
		if (unlikely(rc)) {
			if (DECODING(proc_op))
				D_FREE(p->sg_iovs);
			return rc;
		}
	}
```

Reply:
oh, this function is actually moved here from another place. ok, I can change it to if. sure. thanks.